[SYCL][Graph] Add a shortcut for adding leaves as dependencies (#339)
reble/llvm: sycl/doc/extensions/experimental/sycl_ext_oneapi_graph.asciidoc

    depends_on(NodeTN... nodes);
};

/// Fake node used as a wildcard by users to add all previous graph leaves as
/// dependencies. Note that this wildcard is only handled by the Explicit API.
const sycl::ext::oneapi::experimental::node ALL_LEAVES_SHORTCUT =

>     sycl::detail::createSyclObjFromImpl<sycl::ext::oneapi::experimental::node>(

Reviewer: We shouldn't be exposing anything in the `sycl::detail::` class to an application user. That's a DPC++ implementation detail unspecified by SYCL, and we can't guarantee it won't change or be removed in future.

Author: The implementation update solves this issue.

}
----

> ==== Depends-On-all-leaves Property 

Reviewer: consistency nitpick
```suggestion
==== Depends-On-All-Leaves Property 
```

The API for explicitly adding nodes to a `command_graph` includes a
`property_list` parameter. This extension defines the `depends_on_all_leaves` 
property to be passed here. `depends_on_all_leaves` provides a shortcut for 
adding all leaves of a graph as dependencies.

> Therefore the created node will depend on all the current leaves in the graph.  

Reviewer: Wondering whether we should add some sort of nod to barriers here, e.g "adding an empty node with this property will act as a barrier to future commands"

> adding all leaves of a graph as dependencies.
> Therefore the created node will depend on all the current leaves in the graph.  

Reviewer 2: ```suggestion
adding all leaves of a graph as dependencies of the newly created node.
```

Reviewer 3: Do we need to specify when we actually evaluate the node dependencies? For a multi-threaded scenario the set of leaves wouldn't necessarily be constant.

Author: I merged the two sentences in one but I kept the notion of "current" leaves.

Author: IMO, in a multi-threading scenario this property could lead to an indeterminate graph since the order of adding nodes is not deterministic, but it cannot lead to an invalid state of the graph.

}
----

`ALL_LEAVES_SHORTCUT` provides a shortcut for adding all leaves of a graph as dependencies.
Therefore if this wildcard node is passed to `depends_on` in the explicit API,
the created node will depend on all the current leaves in the graph.  
[source,c++]
----
const sycl::ext::oneapi::experimental::node ALL_LEAVES_SHORTCUT =
    sycl::detail::createSyclObjFromImpl<sycl::ext::oneapi::experimental::node>(
        nullptr);
----

> 

Reviewer 2: I think that trying to achieve this without modifying APIs is admirable, but this feels very hacky to me. If we don't want to modify any existing APIs then perhaps just a new property which will depend on all leaves makes sense? This doesn't change the existing signature of `add()` and is part of the purpose of having an extensible property list.

Author: I updated the implementation using a node property instead of a fake node.
